staging/src/k8s.io/component-helpers/kubernetesx509/kubernetesx509.go

```diff
@@ -0,0 +1,140 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package kubernetesx509 contains routines for creating and parsing
+// x509 certificates and certificate requests that embed Kubernetes-specific
+// X.509 extensions that communicate the Kubernetes identity of a given
+// workload.
+package kubernetesx509
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"fmt"
+)
+
+var (
+	// The ASN.1 Private Enterprise Number assigned to the Kubernetes project.
+	// Used to name X.509 extensions that communicate Kubernetes-specific
+	// concepts.
+	KubernetesPEN = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 57683}
+
+	// OIDPodIdentity identifies the Kubernetes PodIdentity X.509 extension.
+	OIDPodIdentity = makeKubernetesOID(1)
+)
+
+func makeKubernetesOID(subIDs ...int) asn1.ObjectIdentifier {
+	base := asn1.ObjectIdentifier{}
+	base = append(base, KubernetesPEN...)
+	base = append(base, subIDs...)
+	return base
+}
+
+type PodIdentity struct {
+	Namespace          string
+	ServiceAccountName string
+	ServiceAccountUID  string
+	PodName            string
+	PodUID             string
+	NodeName           string
+	NodeUID            string
+}
+
+type podIdentityASN1 struct {
+	Namespace          string `asn1:"utf8"`
+	ServiceAccountName string `asn1:"utf8"`
+	ServiceAccountUID  string `asn1:"utf8"`
+	PodName            string `asn1:"utf8"`
+	PodUID             string `asn1:"utf8"`
+	NodeName           string `asn1:"utf8"`
+	NodeUID            string `asn1:"utf8"`
+}
+
+// AddPodIdentityToCertificate adds a Kubernetes PodIdentity extension to an
+// x509.Certificate.
+//
+// Independent signers can use this routine to add PodIdentity extensions to the
+// certificates they issue.  Note that it is not required, or even expected,
+// that most independent signers will insert these extensions into their issued
+// certificates.
+func AddPodIdentityToCertificate(pod *PodIdentity, template *x509.Certificate) error {
+	podIdentityBytes, err := asn1.Marshal(podIdentityASN1{
+		Namespace:          pod.Namespace,
+		ServiceAccountName: pod.ServiceAccountName,
+		ServiceAccountUID:  pod.ServiceAccountUID,
+		PodName:            pod.PodName,
+		PodUID:             pod.PodUID,
+		NodeName:           pod.NodeName,
+		NodeUID:            pod.NodeUID,
+	})
+	if err != nil {
+		return fmt.Errorf("while ASN1-marshaling PodIdentity extension: %w", err)
+	}
+
+	template.ExtraExtensions = append(template.ExtraExtensions, pkix.Extension{
+		Id:    OIDPodIdentity,
+		Value: podIdentityBytes,
+	})
```

Comment:
marking as critical will populate this into UnhandledCriticalExtensions in go x509 parsing, right?

```go
  // UnhandledCriticalExtensions contains a list of extension IDs that
	// were not (fully) processed when parsing. Verify will fail if this
	// slice is non-empty, unless verification is delegated to an OS
	// library which understands all the critical extensions.
	//
	// Users can access these extensions using Extensions and can remove
	// elements from this slice if they believe that they have been
	// handled.
	UnhandledCriticalExtensions [][asn1](https://pkg.go.dev/encoding/asn1).[ObjectIdentifier](https://pkg.go.dev/encoding/asn1#ObjectIdentifier)
```

what moves this extension out of UnhandledCriticalExtensions so that Verify passes?

I don't necessarily disagree that it should be critical, just trying to understand how that impacts cert verification, etc

Reply:
You are correct --- I will need to change x509.go in order to parse the critical extension and remove it from UnhandledCriticalExtensions before calling Verify on the certificate.

The extension is currently not marked as critical, which is why it's been working in my local testing.